File: src/frontend/src/app/core/services/report.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface FinancialLedgerReport {
  organization_id: number;
  payments: any[];
  total: number;
  total_due: number;
  total_paid: number;
  total_pending: number;
  total_overdue: number;
  collection_rate: number;
  generated_at: string;
}

export interface CollectionSummaryReport {
  organization_id: number;
  collection_rate: number;
  total_due: number;
  total_collected: number;
  total_pending: number;
  total_overdue: number;
  aging_buckets: Record<string, number>;
  monthly_trend: Array<{
    month: string;
    collection_rate: number;
    amount_due: number;
    amount_collected: number;
  }>;
  report_period: string;
  generated_at: string;
}

export interface PaymentAnalysisReport {
  organization_id: number;
  payment_methods: Record<string, number>;
  status_distribution: Record<string, number>;
  daily_trend: Record<string, number>;
  total_payments: number;
  report_period: string;
  generated_at: string;
}

export interface DashboardMetrics {
  collection_summary: CollectionSummaryReport;
  payment_analysis: PaymentAnalysisReport;
  generated_at: string;
}

export interface TenantReportEntry {
  tenant_id: number;
  tenant_name: string;
  phone_number: string;
  email: string;
  unit_number: string;
  building_name: string;
  property_name: string;
  lease_start: string | null;
  lease_end: string | null;
  monthly_rent: number;
  lease_active: boolean;
  total_due: number;
  total_paid: number;
  balance_due: number;
}

export interface TenantSummaryReport {
  organization_id: number;
  tenants: TenantReportEntry[];
  total: number;
  active_tenants: number;
  generated_at: string;
}

export interface PropertyAnalyticsEntry {
  property_id: number;
  property_name: string;
  property_code: string;
  property_type: string;
  payment_stats: Record<string, any> | null;
}

export interface PropertyAnalyticsReport {
  organization_id: number;
  properties: PropertyAnalyticsEntry[];
  total: number;
  report_period: string;
  generated_at: string;
}

@Injectable({
  providedIn: 'root',
})
export class ReportService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/reports`;

  getFinancialLedger(
    page: number = 1,
    pageSize: number = 50,
    filters?: Record<string, any>
  ): Observable<FinancialLedgerReport> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('page_size', pageSize.toString());

    if (filters) {
      Object.entries(filters).forEach(([key, value]) => {
        if (value !== null && value !== undefined && value !== '') {
          params = params.set(key, value.toString());
        }
      });
    }

    return this.http.get<FinancialLedgerReport>(`${this.apiUrl}/ledger`, {
      params,
    });
  }

  getCollectionSummary(startDate?: string, endDate?: string): Observable<CollectionSummaryReport> {
    let params = new HttpParams();

    if (startDate) {
      params = params.set('start_date', startDate);
    }
    if (endDate) {
      params = params.set('end_date', endDate);
    }

    return this.http.get<CollectionSummaryReport>(`${this.apiUrl}/collection-summary`, { params });
  }

  getPaymentAnalysis(startDate?: string, endDate?: string): Observable<PaymentAnalysisReport> {
    let params = new HttpParams();

    if (startDate) {
      params = params.set('start_date', startDate);
    }
    if (endDate) {
      params = params.set('end_date', endDate);
    }

    return this.http.get<PaymentAnalysisReport>(`${this.apiUrl}/payment-analysis`, { params });
  }

  getDashboardMetrics(): Observable<DashboardMetrics> {
    return this.http.get<DashboardMetrics>(`${this.apiUrl}/dashboard-metrics`);
  }

  getTenantSummary(): Observable<TenantSummaryReport> {
    return this.http.get<TenantSummaryReport>(`${this.apiUrl}/tenant-summary`);
  }

  getPropertyAnalytics(startDate?: string, endDate?: string): Observable<PropertyAnalyticsReport> {
    let params = new HttpParams();
    if (startDate) params = params.set('start_date', startDate);
    if (endDate) params = params.set('end_date', endDate);
    return this.http.get<PropertyAnalyticsReport>(`${this.apiUrl}/property-analytics`, { params });
  }
}

File: src/backend/api/internal/services/report_service.go
package services

import (
	"fmt"
	"time"

	"github.com/ysnarafat/tenantly/internal/interfaces"
	"github.com/ysnarafat/tenantly/internal/models"
)

type ReportService struct {
	paymentRepo  interfaces.PaymentRepositoryInterface
	propertyRepo interfaces.PropertyRepositoryInterface
}

func NewReportService(paymentRepo interfaces.PaymentRepositoryInterface, propertyRepo interfaces.PropertyRepositoryInterface) *ReportService {
	return &ReportService{
		paymentRepo:  paymentRepo,
		propertyRepo: propertyRepo,
	}
}

// FinancialLedgerReport returns complete transaction history with balances
func (s *ReportService) FinancialLedgerReport(orgID int, filters map[string]interface{}, limit, offset int) (*models.FinancialLedgerReport, error) {
	filters["organization_id"] = orgID

	payments, total, err := s.paymentRepo.GetWithDetailsAndFilters(filters, limit, offset)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch payment ledger: %w", err)
	}

	summary, err := s.paymentRepo.GetDashboardSummary()
	if err != nil {
		return nil, fmt.Errorf("failed to get summary: %w", err)
	}

	return &models.FinancialLedgerReport{
		OrganizationID: orgID,
		Payments:       payments,
		Total:          total,
		TotalDue:       int64(summary.TotalDue),
		TotalPaid:      int64(summary.TotalPaid),
		TotalPending:   int64(summary.TotalPending),
		TotalOverdue:   int64(summary.TotalOverdue),
		CollectionRate: summary.CollectionRate,
		GeneratedAt:    time.Now(),
	}, nil
}

// CollectionSummaryReport returns collection rates and payment summary
func (s *ReportService) CollectionSummaryReport(orgID int, startDate, endDate time.Time) (*models.CollectionSummaryReport, error) {
	summary, err := s.paymentRepo.GetDashboardSummary()
	if err != nil {
		return nil, fmt.Errorf("failed to get collection summary: %w", err)
	}

	agingBuckets, err := s.paymentRepo.GetAgingBuckets(orgID)
	if err != nil {
		return nil, fmt.Errorf("failed to get aging buckets: %w", err)
	}

	monthlyTrend, err := s.paymentRepo.GetMonthlyCollectionTrend(orgID, 6)
	if err != nil {
		return nil, fmt.Errorf("failed to get monthly trend: %w", err)
	}

	return &models.CollectionSummaryReport{
		OrganizationID: orgID,
		CollectionRate: summary.CollectionRate,
		TotalDue:       int64(summary.TotalDue),
		TotalCollected: int64(summary.TotalPaid),
		TotalPending:   int64(summary.TotalPending),
		TotalOverdue:   int64(summary.TotalOverdue),
		AgingBuckets:   agingBuckets,
		MonthlyTrend:   monthlyTrend,
		ReportPeriod:   fmt.Sprintf("%s to %s", startDate.Format("2006-01-02"), endDate.Format("2006-01-02")),
		GeneratedAt:    time.Now(),
	}, nil
}

// TenantSummaryReport returns all tenants with their lease and payment totals
func (s *ReportService) TenantSummaryReport(orgID int) (*models.TenantSummaryReport, error) {
	entries, err := s.paymentRepo.GetTenantPaymentSummary(orgID)
	if err != nil {
		return nil, fmt.Errorf("failed to get tenant summary: %w", err)
	}

	active := 0
	for _, e := range entries {
		if e.LeaseActive {
			active++
		}
	}

	return &models.TenantSummaryReport{
		OrganizationID: orgID,
		Tenants:        entries,
		Total:          len(entries),
		ActiveTenants:  active,
		GeneratedAt:    time.Now(),
	}, nil
}

// PropertyAnalyticsReport returns per-property payment analytics for the organisation
func (s *ReportService) PropertyAnalyticsReport(orgID int, startDate, endDate time.Time) (*models.PropertyAnalyticsReport, error) {
	properties, _, err := s.propertyRepo.List(map[string]interface{}{
		"organization_id": orgID,
		"active":          true,
	}, 200, 0)
	if err != nil {
		return nil, fmt.Errorf("failed to list properties: %w", err)
	}

	entries := make([]*models.PropertyAnalyticsEntry, 0, len(properties))
	for _, p := range properties {
		stats, err := s.paymentRepo.GetPropertyPaymentStats(p.ID, startDate, endDate)
		if err != nil {
			stats = nil
		}
		entries = append(entries, &models.PropertyAnalyticsEntry{
			PropertyID:   p.ID,
			PropertyName: p.PropertyName,
			PropertyCode: p.PropertyCode,
			PropertyType: string(p.PropertyType),
			PaymentStats: stats,
		})
	}

	return &models.PropertyAnalyticsReport{
		OrganizationID: orgID,
		Properties:     entries,
		Total:          len(entries),
		ReportPeriod:   fmt.Sprintf("%s to %s", startDate.Format("2006-01-02"), endDate.Format("2006-01-02")),
		GeneratedAt:    time.Now(),
	}, nil
}

// PaymentAnalysisReport returns payment methods and status distribution
func (s *ReportService) PaymentAnalysisReport(orgID int, startDate, endDate time.Time) (*models.PaymentAnalysisReport, error) {
	filters := map[string]interface{}{
		"organization_id": orgID,
	}

	payments, _, err := s.paymentRepo.GetWithDetailsAndFilters(filters, 10000, 0)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch payments: %w", err)
	}

	methodCounts := make(map[string]int)
	statusCounts := make(map[string]int64)
	dailyTrend := make(map[string]int64)

	if payments != nil {
		for _, p := range payments {
			if p.CreatedAt.After(startDate) && p.CreatedAt.Before(endDate) {
				method := "cash"
				if p.PaymentMethod != "" {
					method = p.PaymentMethod
				}
				methodCounts[method]++
				statusCounts[string(p.Status)]++
				day := p.CreatedAt.Format("2006-01-02")
				dailyTrend[day]++
			}
		}
	}

	return &models.PaymentAnalysisReport{
		OrganizationID:     orgID,
		PaymentMethods:     methodCounts,
		StatusDistribution: statusCounts,
		DailyTrend:         dailyTrend,
		TotalPayments:      int64(len(payments)),
		ReportPeriod:       fmt.Sprintf("%s to %s", startDate.Format("2006-01-02"), endDate.Format("2006-01-02")),
		GeneratedAt:        time.Now(),
	}, nil
}
